### PR TITLE
added multicommand functionality to the Rasa-Vim-Ext

### PR DESCRIPTION
**Pull request by robinvd**
Changelist:

added VimHist data type and helper functions
changed HandleKeypress , normal and insert functions to use the new data type
changed the 'f' and 'F' functions to use the new multicommand functionality
added the 't' and 'T' commands
changed setStatus so that the pending character is shown in the statusbar
changed Esc so it moves back one char

### CODE REVIEW

File: rasa-ext-vim/src/Rasa/Ext/Vim.hs [hunk 1]
```diff
@@ -52,17 +72,23 @@ handleKeypress :: Keypress -> Action ()
 handleKeypress keypress = do
   focMode <- focusDo $ do
     mode <- getVim
+    VimHist hist <- getHist
     case mode of
-      Normal -> normal keypress
-      Insert -> insert keypress
+      Normal -> normal $  hist ++ [keypress]
```
Comment (ChrisPenner) [inline rasa-ext-vim/src/Rasa/Ext/Vim.hs]: I like this better, the list isn't ever going to be longer than like 3 keypresses so performance isn't an issue.

**Comment by robinvd** [comment]
I agree that it is not the best solution as it tries to pattern match all options evertime a button is pressed (or is this optimized by ghc). the best would be a tree of actions i think, after a Keypress a subtree would be selected as remaining possible options. I'll look at the things you mentioned. Thanks for looking at it.

File: rasa-ext-vim/src/Rasa/Ext/Vim.hs [hunk 2]
```diff
@@ -72,59 +98,66 @@ global _ (Keypress 'c' [Ctrl]) = exit
...
-normal (Keypress 'F' _) = findPrev "f"
-normal (Keypress 'X' _) = moveRangesByN (-1) >> delete
-normal (Keypress 'x' _) = delete
-normal (Keypress 's' [Ctrl]) = save
-normal (Keypress ';' _) = ranges <~ use (ranges.reversed.to (take 1))
+normal [Keypress 'f' _,Keypress x []] = findNext $ T.singleton x
+normal [Keypress 'f' _] = addHist $ Keypress 'f' []
+normal [Keypress 't' _,Keypress x []] = findNext (T.singleton x) >> moveRangesByN (-1)
+normal [Keypress 't' _] = addHist $ Keypress 't' []
+normal [Keypress 'T' _,Keypress x []] = findNext (T.singleton x) >> moveRangesByN (-1)
```
Comment (robinvd) [inline rasa-ext-vim/src/Rasa/Ext/Vim.hs]: By backwards you mean the addHist case first?

File: rasa-ext-vim/src/Rasa/Ext/Vim.hs [hunk 3]
```diff
@@ -33,6 +39,20 @@ getVim = use bufExt
 setMode :: VimSt -> BufAction ()
 setMode vimst = bufExt .= vimst
 
+-- | Same helper fuctions as VimSt but for VimHist
```
Comment (robinvd) [inline rasa-ext-vim/src/Rasa/Ext/Vim.hs]: Changed it for VimSt as well as VimHist

File: rasa-ext-vim/src/Rasa/Ext/Vim.hs (hunk 2 above)
Comment (robinvd) [inline rasa-ext-vim/src/Rasa/Ext/Vim.hs]: Think i changed what you meant

**Comment by ChrisPenner** [comment]
Okay looks good! 

For future Pull-requests it's convenient if you can Squash your commits once you've got the feature finished; there's a guide here: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/ 

Not going to let it hold this one up though, thanks for the help!

**Comment by ChrisPenner** [comment]
What I'd meant by the comment on the `T` comment is that right now `T` searches forwards the same as `t` haha, don't worry though I can switch it over;

**Comment by ChrisPenner** [comment]
For your own reference, there's a few tricks you can do with the lenses that can help out; I did a small amount of re-structuring to use them inline. You can remove the need for pattern-matching on VimHist for example, you can see the results here: https://github.com/ChrisPenner/rasa/blob/master/rasa-ext-vim/src/Rasa/Ext/Vim.hs#L40

Overall this works great, makes it feel a lot more like vim!

**Pull request merged**